components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,343 @@
+    }
+
+    let builder = SuggestionBuilder {
+        query,
+        subject: &subject_candidate,
+        pre_modifier,
+        post_modifier,
+        location_sign,
+        location,
+        need_location,
+    };
+    Ok(Some(builder.into()))
+}
+
+fn find_location(
+    conn: &Connection,
+    query: &str,
+) -> Result<(String, Option<String>, Option<String>, bool)> {
+    let query_with_spaces = format!(" {} ", query.trim());
+    let results: Vec<(usize, usize, i8)> = conn.query_rows_and_then(
+        "
+        SELECT
+          INSTR(:query, ' ' || keyword || ' ') AS sign_index,
+          LENGTH(keyword) AS sign_length,
+          need_location
+        FROM yelp_location_signs
+        WHERE sign_index ORDER BY sign_length DESC LIMIT 1
+        ",
+        named_params! {
+            ":query": &query_with_spaces.to_lowercase(),
+        },
+        |row| -> Result<_> {
+            Ok((
+                row.get::<_, usize>("sign_index")?,
+                row.get::<_, usize>("sign_length")?,
+                row.get::<_, i8>("need_location")?,
+            ))
+        },
+    )?;
+
+    if results.is_empty() {
+        Ok((query.trim().to_string(), None, None, false))
+    } else {
+        let (sign_index, sign_length, need_location) = &results[0];
+        let pre_location = query_with_spaces.get(..*sign_index).unwrap().trim();
+        let location_sign = query_with_spaces
+            .get(*sign_index..*sign_index + *sign_length)
+            .unwrap()
+            .trim();
+        let location = query_with_spaces
+            .get(*sign_index + *sign_length..)
+            .unwrap()
+            .trim();
+        Ok((
+            pre_location.to_string(),
+            if location_sign.is_empty() {
+                None
+            } else {
+                Some(location_sign.to_string())
+            },
+            if location.is_empty() {
+                None
+            } else {
+                Some(location.to_string())
+            },
+            *need_location == 1,
+        ))
+    }
+}
+
+fn find_modifiers(
+    conn: &Connection,
+    query: &str,
+) -> Result<(String, Option<String>, Option<String>)> {
+    if !query.contains(" ") {
+        return Ok((query.to_string(), None, None));
+    }
+
+    let words: Vec<_> = query.split_whitespace().collect();
+
+    let mut pre_modifier = None;
+    for i in 0..MAX_MODIFIER_WORDS_LENGTH {
+        let mut candidate_chunks = words.chunks(MAX_MODIFIER_WORDS_LENGTH - i);
+        let candidate = candidate_chunks.next().unwrap().join(" ");
+        // Search from longer term.
+        if is_modifier(conn, &candidate, Modifier::PreModifier)? {
+            pre_modifier = Some(candidate);
+            break;
+        }
+    }
+
+    let mut post_modifier = None;
+    for i in 0..MAX_MODIFIER_WORDS_LENGTH {
+        let mut candidate_chunks = words.rchunks(MAX_MODIFIER_WORDS_LENGTH - i);
+        let candidate = candidate_chunks.next().unwrap().join(" ");
```

Comment:
Likewise, this would only check the _last_ chunk for post-modifiers. Did you mean to only check the first two words for pre-modifier(s) and the last two words for post-modifier(s)? If so, I'd suggest calling `candidate_chunks.nth(0)` to make it clearer.